unittests/test-contracts/if_ibc/if_ibc.hpp

```diff
@@ -0,0 +1,489 @@
+#include <eosio/eosio.hpp>
+#include <eosio/crypto.hpp>
+#include <eosio/crypto_bls_ext.hpp>
+#include <eosio/system.hpp>
+#include <eosio/transaction.hpp>
+
+#include "bitset.hpp"
+
+#include <numeric>
+#include <queue>
+
+using namespace eosio;
+
+CONTRACT if_ibc : public contract {
+   public:
+      using contract::contract;
+
+      using bls_public_key = std::vector<char>;
+      using bls_signature = std::vector<char>;
+
+      const uint32_t POLICY_CACHE_EXPIRY = 600; //10 minutes for testing
+      const uint32_t PROOF_CACHE_EXPIRY = 600; //10 minutes for testing
+
+      //Compute the next power of 2 for a given number
+      static uint64_t next_power_of_2(uint64_t value) {
+          value -= 1;
+          value |= value >> 1;
+          value |= value >> 2;
+          value |= value >> 4;
+          value |= value >> 8;
+          value |= value >> 16;
+          value |= value >> 32;
+          value += 1;   return value;
+      }
+
+      static uint64_t clz_power_of_2(uint64_t v) {
+          return 64 - (__builtin_clzll(v)+1); 
+      }
+
+      //Compute the maximum number of layers of a merkle tree for a given number of leaves
+      static uint64_t calculate_max_depth(uint64_t node_count) {
+          if (node_count == 0) return 0;
+          auto implied_count = next_power_of_2(node_count);
+          return clz_power_of_2(implied_count) + 1;
+      }

```

Comment:
I was intrigued by the weird function name `clz_power_of_2`.
What that function does is more like  `log_base_2`

I think this could be replaced by:
```suggestion
      //Compute the maximum number of layers of a merkle tree for a given number of leaves
      static uint64_t calculate_max_depth(uint64_t node_count) {
          if (node_count == 0) return 0;
          return std::llround(std::log2(node_count)) + 2;
      }
```

Reply:
That's what is, yes.

Your suggestion seems to work fine, so I adjusted that.

Originally, I ported the code that was here : 

https://github.com/AntelopeIO/leap/blob/hotstuff_integration/libraries/chain/include/eosio/chain/incremental_merkle.hpp#L59

When Bucky reviewed the contract, he told me about this __builtin_clzll function.

Maybe we want to fix that code as well?

Reply:
> Maybe we want to fix that code as well?

Thanks, sounds good, let me do some verification first and I'll create a PR for it.

Reply:
@systemzax, if you need a function which provides exactly the same number as before, please use:

```
static uint64_t calculate_max_depth(uint64_t node_count) {
   if (node_count == 0)
      return 0;
   return std::llround(std::ceil(std::log2(node_count)))+1ul;
}
```

Reply:
@systemzax Matt came up with a faster version, but it requires C++20 which I don't believe is supported in cdt
(See PR #2320)

Reply:
the non c++20 way would be something like
```c++
uint64_t calculate_max_depth(uint64_t node_count) {
   if(node_count <= 1)
      return node_count;
   return 64 - __builtin_clzll(2 << (64 - 1 - __builtin_clzll ((node_count - 1))));
}
```
but this doesn't optimize quite as nicely as the c++20 variant for some reason (c++20 variant only uses a single `lzcnt` instruction instead of two as the above does). I'm not sure off hand how to improve it. Still, webassembly for the above compiles down to
```wat
calculate_max_depthr(unsigned long long):              # @calculate_max_depthr(unsigned long long)
        block           
        local.get       0
        i64.const       2
        i64.lt_u
        br_if           0                               # 0: down to label0
        i64.const       64
        i32.const       2
        i32.const       63
        local.get       0
        i64.const       -1
        i64.add 
        i64.clz 
        i32.wrap_i64
        i32.sub 
        i32.shl 
        i64.extend_i32_s
        i64.clz 
        i64.sub 
        local.set       0
        end_block                               # label0:
        local.get       0
        end_function
```
which is pretty good

Reply:
This version looks good, same result as prior version for all values except for `1`, where it returns `2` instead of `1`. I think it is probably fine to use.

Reply:
> except for `1`, where it returns `2` instead of `1`

Maybe you didn't copy the entirety of the function, it starts with
```cpp
   if(node_count <= 1)
      return node_count;
```
so idk how an input of `1` could return `2`

Reply:
> Maybe you didn't copy the entirety of the function,

Right, my bad, thanks for the correction Matt.

@systemzax Matt's version is exactly equivalent (and simpler) than the original one, so I think it would be best to adopt it.